### PR TITLE
[SPARK-48715][SQL] Integrate UTF8String validation into collation-aware string function implementations

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Use our own invalid UTF-8 byte sequence replacement logic in UTF8String, before all `.toString()` method calls.

### Why are the changes needed?
Avoid relying on Java to perform invalid UTF-8 byte sequence replacement, and ensure consistent results.

### Does this PR introduce _any_ user-facing change?
Yes, collation aware string function implementations will now rely on our own invalid UTF-8 string replacement implementation, instead of Java's.

### How was this patch tested?
Existing tests, with some changes in `UTF8StringSuite` and `CollationSupportSuite`.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -329,6 +343,27 @@ private static void appendReplacementCharacter(ArrayList<Byte> bytes) {
    * @return A new UTF8String that is a valid UTF8 string.
    */
   public UTF8String makeValid() {
+    return UTF8String.fromBytes(makeValidBytes());
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:346]:
```suggestion
    if (isValid) return this;
    return UTF8String.fromBytes(makeValidBytes());
```
I believe we can avoid recomputing validity if we already know the current UTF8 string is valid.

**Review by uros-db — COMMENTED** [review]
@mkaravel ready for another round of review

**Review by mkaravel — APPROVED** [review]
LGTM

**Review by uros-db — COMMENTED** [review]
@cloud-fan ready for review

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**